GVFS/GVFS.Common/ProductUpgrader.cs

```diff
@@ -0,0 +1,215 @@
+            Exception exception;
+            if (!this.fileSystem.TryCreateDirectory(toolsDirectoryPath, out exception))
+            {
+                upgraderToolPath = null;
+                error = exception.Message;
+                this.TraceException(
+                    exception,
+                    nameof(this.TrySetupToolsDirectory),
+                    $"Error creating upgrade tools directory {toolsDirectoryPath}.");
+                return false;
+            }
+
+            string currentPath = ProcessHelper.GetCurrentProcessLocation();
+            error = null;
+            foreach (string name in UpgraderToolAndLibs)
+            {
+                string toolPath = Path.Combine(currentPath, name);
+                string destinationPath = Path.Combine(toolsDirectoryPath, name);
+                try
+                {
+                    this.fileSystem.CopyFile(toolPath, destinationPath, overwrite: true);
+                }
+                catch (UnauthorizedAccessException e)
+                {
+                    error = string.Join(
+                        Environment.NewLine,
+                        "File copy error - " + e.Message,
+                        $"Make sure you have write permissions to directory {toolsDirectoryPath} and run {GVFSConstants.UpgradeVerbMessages.GVFSUpgradeConfirm} again.");
+                    this.TraceException(e, nameof(this.TrySetupToolsDirectory), $"Error copying {toolPath} to {destinationPath}.");
+                    break;
+                }
+                catch (IOException e)
+                {
+                    error = "File copy error - " + e.Message;
+                    this.TraceException(e, nameof(this.TrySetupToolsDirectory), $"Error copying {toolPath} to {destinationPath}.");
+                    break;
+                }
+            }
+
+            if (string.IsNullOrEmpty(error))
+            {
+                // There was no error - set upgradeToolPath and return success.
+                upgraderToolPath = Path.Combine(toolsDirectoryPath, UpgraderToolName);
+                return true;
+            }
+            else
+            {
+                // Encountered error - do not set upgrade tool path and return failure.
+                upgraderToolPath = null;
+                return false;
+            }
+        }
+
+        public abstract bool TryCleanup(out string error);
+
+        public void TraceException(Exception exception, string method, string message)
+        {
+            this.TraceException(this.tracer, exception, method, message);
+        }
+
+        public void TraceException(ITracer tracer, Exception exception, string method, string message)
+        {
+            EventMetadata metadata = new EventMetadata();
+            metadata.Add("Method", method);
+            metadata.Add("Exception", exception.ToString());
+            tracer.RelatedError(metadata, message, Keywords.Telemetry);
```

Comment:
It's not hurting anything by including it, but all `ReleatedError(...)` calls are recorded to telemetry automatically:

```
        public virtual void RelatedError(EventMetadata metadata, string message)
        {
            this.RelatedError(metadata, message, Keywords.Telemetry);
        }
```